File: frontend/src/utils/metaPeriodo.ts
/** Helpers do toggle Bruto/Líquido e Configuração do Período (Dashboard). */

export type VisaoReceita = 'liquido' | 'bruto';

export function metaBruta(
  metaLiquida: number | null | undefined,
  aliquotaPeriodo: number | null | undefined,
): number | null {
  if (metaLiquida == null || aliquotaPeriodo == null) return null;
  const liq = Number(metaLiquida);
  const aliq = Number(aliquotaPeriodo);
  if (!Number.isFinite(liq) || !Number.isFinite(aliq)) return null;
  if (aliq < 0 || aliq >= 100) return null;
  const denom = 1 - aliq / 100;
  if (denom <= 0) return null;
  return Math.round((liq / denom) * 100) / 100;
}

export function rotuloVisao(visao: VisaoReceita): string {
  return visao === 'bruto' ? 'visão bruta' : 'visão líquida';
}

export function valorPorVisao(
  liquido: number | null | undefined,
  bruto: number | null | undefined,
  visao: VisaoReceita,
): number {
  const v = visao === 'bruto' ? bruto : liquido;
  return Number(v) || 0;
}

export function pctPorVisao(
  pctLiquido: number | null | undefined,
  pctBruto: number | null | undefined,
  visao: VisaoReceita,
): number | null {
  const p = visao === 'bruto' ? pctBruto : pctLiquido;
  if (p == null || !Number.isFinite(p)) return null;
  return p;
}

File: backend/app/services/meta_periodo.py
"""Helpers da Configuração do Período (meta líquida + alíquota)."""
from __future__ import annotations


def meta_bruta(meta_liquida: float | None, aliquota_periodo: float | None) -> float | None:
    """meta_bruta = round(meta_liquida / (1 - aliquota/100), 2). None se inválido."""
    if meta_liquida is None or aliquota_periodo is None:
        return None
    try:
        liq = float(meta_liquida)
        aliq = float(aliquota_periodo)
    except (TypeError, ValueError):
        return None
    if aliq < 0 or aliq >= 100:
        return None
    denom = 1.0 - (aliq / 100.0)
    if denom <= 0:
        return None
    return round(liq / denom, 2)


def aliquotas_iguais(a: float | None, b: float | None) -> bool:
    if a is None and b is None:
        return True
    if a is None or b is None:
        return False
    return abs(float(a) - float(b)) < 1e-9
